### PR TITLE
ZNTA-699 navigation flow migration

### PR DESCRIPTION
**Pull request by huangp**
https://zanata.atlassian.net/browse/ZNTA-699

Note that this is based on (and it targets) the branch in https://github.com/zanata/zanata-server/pull/980

### CODE REVIEW

**Comment by huangp** [comment]
Sean has reviewed it in #959 Need another JSF expert to review again here

File: zanata-war/src/main/java/org/zanata/action/LoginAction.java [hunk 1]
```diff
@@ -105,9 +116,35 @@ public String login() {
...
+                return "dashboard";
+            }
+            if (authenticationManager.isAuthenticated() && !authenticationManager.isNewUser() && userRedirect.isRedirect()) {
+                // TODO [CDI] seam will create a conversation when you return view id directly or redirect to external url
+                return continueToPreviousUrl();
+            }
+        } else if ("inactive".equals(loginResult)) {
+            Conversation.instance().end();
+            return "inactive";
+        }
```
Comment (carlosmunoz) [inline zanata-war/src/main/java/org/zanata/action/LoginAction.java]: I imagine all this is what currently sits on pages.xml, it was simply moved to java.

File: zanata-war/src/main/java/org/zanata/action/MergeTransAction.java [hunk 2]
```diff
@@ -184,6 +186,9 @@ public void startMergeTranslations() {
         mergeTranslationsManager.start(sourceProjectSlug,
             sourceVersionSlug, targetProjectSlug, targetVersionSlug,
             !keepExistingTranslation);
+        FacesContext context = FacesContext.getCurrentInstance();
+        NavigationHandler navigationHandler = context.getApplication().getNavigationHandler();
+        navigationHandler.handleNavigation(context, null, "merge-translation");
```
Comment (carlosmunoz) [inline zanata-war/src/main/java/org/zanata/action/MergeTransAction.java]: Why do we need this? Shouldn't this be going to the same page where the action is triggered by default?
Also, sounds like it's something that would commonly be used so maybe we should put it in some common class.

File: zanata-war/src/main/java/org/zanata/action/ProjectHome.java [hunk 3]
```diff
@@ -286,8 +286,6 @@ public void setSelectedProjectType(String selectedProjectType) {
                 && !selectedProjectType.equals("null")) {
             ProjectType projectType = ProjectType.valueOf(selectedProjectType);
             getInstance().setDefaultProjectType(projectType);
-        } else {
-            getInstance().setDefaultProjectType(null);
```
Comment (carlosmunoz) [inline zanata-war/src/main/java/org/zanata/action/ProjectHome.java]: This looks like it doesn't have much to do with navigation. Why did we remove these lines?

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml [hunk 4]
```diff
@@ -78,4 +78,387 @@
...
+      <to-view-id>#{userRedirect.url}</to-view-id>
+      <redirect/>
+    </navigation-case>
+    <navigation-case>
+      <!-- Default redirection after a log in is to the dashboard -->
+      <if>#{identity.loggedIn}</if>
+      <to-view-id>/dashboard/home.xhtml</to-view-id>
+      <redirect/>
+    </navigation-case>
+  </navigation-rule>
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: On a previous comment I noted that one of the actions (`LoginAction`) had new Java code to determine navigation. Seems like this is the same thing but in `faces-config.xml`. Should we standardize on one place?

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml [hunk 5]
```diff
@@ -78,4 +78,387 @@
...
+      &lt;!&ndash; TODO [CDI] end conversation (it will lose notification message) &ndash;&gt;
+      &lt;!&ndash;<end-conversation />&ndash;&gt;
+      <redirect/>
+    </navigation-case>
+  </navigation-rule>-->
+
+  <navigation-rule>
+    <from-view-id>/account/register.xhtml</from-view-id>
+    <navigation-case>
+      <from-outcome>true</from-outcome>
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: Should this be `success`, as is being used on several of the Actions?

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml [hunk 6]
```diff
@@ -78,4 +78,387 @@
...
+      <to-view-id>/dashboard/home.xhtml</to-view-id>
+      <redirect/>
+    </navigation-case>
+  </navigation-rule>
+
+  <!-- Project -->
+  <navigation-rule>
+    <from-view-id>/project/project.xhtml</from-view-id>
+    <navigation-case>
+      <from-outcome>projectSlugUpdated</from-outcome>
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: Some other places are using dashes. What about `project-slug-updated`... or conversely switch everything else to camel case. I think this is the odd one out.

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml [hunk 7]
```diff
@@ -78,4 +78,387 @@
...
+        </view-param>
+        <view-param>
+          <name>iterationSlug</name>
+          <value>#{versionHome.instance.slug}</value>
+        </view-param>
+      </redirect>
+    </navigation-case>
+    <navigation-case>
+      <from-action>#{versionHome.update}</from-action>
+      <from-outcome>versionSlugUpdated</from-outcome>
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: another camel case vs dashed scenario

File: zanata-war/src/main/webapp/WEB-INF/layout/version/merge_trans_modal.xhtml [hunk 8]
```diff
@@ -158,7 +158,7 @@
               action="#{mergeTransAction.startMergeTranslations}"
               value="#{msgs['jsf.iteration.mergeTrans']}"/>
 
-            <ui:fragment rendered="#{mergeTransAction.isCopyActionsRunning(mergeTransAction.targetProjectSlug, mergeTransAction.targetVersionSlug)}">
+            <ui:fragment rendered="#{mergeTransAction.isCopyActionsRunning()}">
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/layout/version/merge_trans_modal.xhtml]: :+1: Good find, but this looks like it's fixing a different issue. Not sure if we should wait for CDI to also fix this.

File: zanata-war/src/main/webapp/WEB-INF/pages.xml [hunk 9]
```diff
@@ -111,14 +14,7 @@
...
 
-  <page view-id="/account/register.xhtml">
-    <navigation from-action="#{identity.isLoggedIn()}">
-      <rule if-outcome="true">
-        <redirect view-id="/dashboard/home.xhtml" />
-      </rule>
-    </navigation>
-  </page>
-
+  <!-- TODO [CDI] uses conversation scope facesMessages. if migrated to faces-config, it will lose notification message -->
```
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/pages.xml]: Do we already have a solution for when we are on CDI?

File: zanata-war/src/main/java/org/zanata/action/LoginAction.java (hunk 1 above)
Comment (huangp) [inline zanata-war/src/main/java/org/zanata/action/LoginAction.java]: Yes. From memory faces-config has less functionality than pages.xml. e.g. it seems only work with string result not just from different EL.

File: zanata-war/src/main/java/org/zanata/action/MergeTransAction.java (hunk 2 above)
Comment (huangp) [inline zanata-war/src/main/java/org/zanata/action/MergeTransAction.java]: This is in an AJAX context not a normal form action. I think this is the only place that requires this at the moment.

Comment (carlosmunoz) [inline zanata-war/src/main/java/org/zanata/action/MergeTransAction.java]: Ahh... so it won't apply the faces-config navigation rule from an ajax request to an action?

File: zanata-war/src/main/java/org/zanata/action/ProjectHome.java (hunk 3 above)
Comment (huangp) [inline zanata-war/src/main/java/org/zanata/action/ProjectHome.java]: When I test it after the change, upon opening up project home, one of the request (I think it's one of the ajax request but hard to verify) keep call this method and set the project type to null. This method is meant to be called only if user changes project type and there is no way they can select null. So I remove it.

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml (hunk 5 above)
Comment (huangp) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: This is a result from ```#{identity.isLoggedIn}```

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml (hunk 6 above)
Comment (huangp) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: It was using this value in pages.xml. I am just copying it over. I will make a new commit to change the name.

File: zanata-war/src/main/webapp/WEB-INF/pages.xml (hunk 9 above)
Comment (huangp) [inline zanata-war/src/main/webapp/WEB-INF/pages.xml]: JSF's flash scope in theory is designed to solve this problem (post -> redirect -> show the message). Sean thinks the deltaspike ViewAccessScope (as a replacement for conversation scope) may work.

Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/pages.xml]: I think flash scope is a better option. We could do this right now if we wanted to. Something like:
```java
FacesContext.getCurrentInstance().getExternalContext().getFlash()
```

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml (hunk 5 above)
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: I think we should change it to a string. The navigation case is too general and makes it hard to understand what happens. I think in general (i.e. whenever possible) we should use the `<from-action>` parameter to be able to clearly understand how it's navigating.

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml (hunk 6 above)
Comment (carlosmunoz) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: Ok cool

File: zanata-war/src/main/webapp/WEB-INF/faces-config.xml (hunk 5 above)
Comment (huangp) [inline zanata-war/src/main/webapp/WEB-INF/faces-config.xml]: Actually this case won't work. The action is invoked during preRenderView, I have to use explicit redirect to make it work. Now the navigation rule is removed

**Comment by huangp** [comment]
Recreated in https://github.com/zanata/zanata-server/pull/1001 to target integration/master

**Pull request closed without merging**